src/bio/bio_xstream.c

```diff
@@ -71,253 +71,28 @@ struct bio_nvme_data {
-			 *  string is null terminated.
-			 */
-			addr_split[0] = ptr[0];
-			addr_split[1] = '\0';
-			strcat(vmd_addr, addr_split);
-			break;
-		}
-		strcat(vmd_addr, ":");
-		ptr = ptr + 2;
-		iteration++;
-	}
-	n = snprintf(dst, SPDK_NVMF_TRADDR_MAX_LEN, "%s", vmd_addr);
-	if (n < 0 || n > SPDK_NVMF_TRADDR_MAX_LEN) {
-		D_ERROR("snprintf failed\n");
-		return -DER_INVAL;
-	}
-
-	return 0;
-}
-
-static int
-populate_whitelist(struct spdk_env_opts *opts)
-{
-	struct spdk_nvme_transport_id	*trid;
-	struct spdk_conf_section	*sp;
-	const char			*val;
-	size_t				 i;
-	int				 rc = 0;
-	bool				 vmd_enabled = false;
-
-	/* Don't need to pass whitelist for non-NVMe devices */
-	if (nvme_glb.bd_bdev_class != BDEV_CLASS_NVME)
-		return 0;
-
-	/*
-	 * Optionally VMD devices will be used, and will require a different
-	 * transport id to pass to whitelist for DPDK.
-	 */
-	sp = spdk_conf_find_section(NULL, "Vmd");
-	if (sp != NULL) {
-		if (spdk_conf_section_get_boolval(sp, "Enable", false))
-			vmd_enabled = true;
-	}
-
-	sp = spdk_conf_find_section(NULL, "Nvme");
-	if (sp == NULL) {
-		D_ERROR("unexpected empty config\n");
-		return -DER_INVAL;
-	}
-
-	D_ALLOC_PTR(trid);
-	if (trid == NULL)
-		return -DER_NOMEM;
-
-	for (i = 0; i < DAOS_NVME_MAX_CTRLRS; i++) {
-		val = spdk_conf_section_get_nmval(sp, "TransportID", i, 0);
-		if (val == NULL) {
-			break;
-		}
-
-		rc = spdk_nvme_transport_id_parse(trid, val);
-		if (rc < 0) {
-			D_ERROR("Unable to parse TransportID: %s\n", val);
-			rc = -DER_INVAL;
-			break;
-		}
-
-		if (trid->trtype != SPDK_NVME_TRANSPORT_PCIE) {
-			D_ERROR("unexpected non-PCIE transport\n");
-			rc = -DER_INVAL;
-			break;
-		}
-
-		if (vmd_enabled) {
-			if (strncmp(trid->traddr, "0", 1) != 0) {
-				/*
-				 * We can assume this is the transport id of the
-				 * backing NVMe SSD behind the VMD. DPDK will
-				 * not recognize this transport ID, instead need
-				 * to pass VMD address as the whitelist param.
-				 */
-				rc = traddr_to_vmd(trid->traddr, trid->traddr);
-				if (rc < 0) {
-					D_ERROR("Invalid traddr=%s\n",
-						trid->traddr);
-					rc = -DER_INVAL;
-					break;
-				}
-			}
-		}
-
-		rc = opts_add_pci_addr(opts, &opts->pci_whitelist,
-				       trid->traddr);
-		if (rc < 0) {
-			D_ERROR("Invalid traddr=%s\n", trid->traddr);
-			rc = -DER_INVAL;
-			break;
-		}
-
-		/* Clear it for the next loop */
-		memset(trid, 0, sizeof(*trid));
-	}
-
-	D_FREE(trid);
-	if (rc && opts->pci_whitelist != NULL) {
-		D_FREE(opts->pci_whitelist);
-		opts->pci_whitelist = NULL;
-	}
-
-	return rc;
-}
-
 static int
 bio_spdk_env_init(void)
 {
 	struct spdk_env_opts	 opts;
 	int			 rc;
 
 	D_ASSERT(nvme_glb.bd_nvme_conf != NULL);
-	if (spdk_conf_first_section(nvme_glb.bd_nvme_conf) == NULL) {
-		D_ERROR("Invalid NVMe conf format\n");
-		return -DER_INVAL;
-	}
-
-	spdk_conf_set_as_default(nvme_glb.bd_nvme_conf);
 
 	spdk_env_opts_init(&opts);
 	opts.name = "daos";
+	opts.shm_id = getpid();
```

Comment:
I remembered that in old version SPDK, it'll generate a 'shm_id' named file in the user's HOME directory, so that later start process with same 'shm_id' with be able to know that a primary process has started (by checking if the existence of the file) and failed to run in primary mode with same 'shm_id'.

Above mechanism will be broken when we start servers on multiple nodes with NFS mounted home directory (What if the engines on two nodes happen to have the same pid?).

I'm not sure if this has been changed in newer SPDK version, but just wondering why we have to set this 'shm_id' now?

Reply:
don't think we need this any more, it was introduced to avoid the resource in use messages with regard to VFIO device files being opened by multiple engines, I will remove